Adding UserReportDoc to database model.

JCantu248 opened the pull request:
Adding UserReportDocs to the PCA database.

## 🗣 Description ##

A document has been added to the PCA database to track when customers reply to campaign emails.

A new command `pca-import --user-report`  was added to allow UserReportDocs to be brought into the PCA database.

A new command `pca-output --output-user-report`  was added to allow a user to output UserReportDocs for a given customer and assessment. 

## 💭 Motivation and context ##

Kelly Thiele and the PCA team want to be able to track when a customer responds to a phishing email generated by the PCA team. They want to store records of when the customer first responded to their campaign, and how many times they have done so. This required changes to the database model which have been made here.
## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

These changes were tested by using the `pca-import --user-report` and `pca-tool --output-user-report ` commands. 

I searched the development PCA database to find corresponding customer ids, assessment ids, and campaign ids. For security purposes the exact data cannot be shown. I used that data to build a JSON object of the following design: 

```
{
	"customer":  ,
	"assessment": ,
	"campaign": 
	"first_report": ,
	"total_num_reports": 

} 
```

I passed the JSON file to the importer command, and then used the `pca-tool --output-user-report` command to view the loaded data. This demonstrates the correct operation of reading and writing to the database using the new document. 

I tested the importer with malformed JSON input, and tested it against previously loaded data to ensure it would block the attempt without the --force flag. 

The output command was tested to ensure correct data was displayed, according to the provided input. 

Made some changes to the code and retested to make sure that when forcing records into the database, it would rewrite over the previously entered data. 

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] All new and existing tests pass.

JCantu248 committed:
Adding UserReportDoc to database model. This changes is requested by Kelly Thiele to be able to track when an end user responds to phishing emails from assessments generated by the team.

jsf9k commented on pca/db/database_model.py:
```diff
@@ -84,7 +86,7 @@ class Meta:
         final = True  # so we don't get a '_cls' field in these documents (e.g. "_cls" : "__main__.Customer")
 
-    def get_all_customers():
+    def get_all_customers(self):
```
Is this change necessary?  Can this method remain static?

JCantu248 replied:
Yes, that can remain static. I'll make the change.

jsf9k reviewed (commented):
I have one minor question.

dav3r commented:
@JCantu248 - Can you please fill out the "Testing" section in the PR description so that we can understand how you tested these changes?

Also, please take care of all of the items in the "Checklist" section.  Ideally, those should all be checked before requesting a PR review.  If any of them do not apply to this particular PR, you can delete them (just don't delete items from that list because you don't feel like doing them).

JCantu248 committed:
Reverted get_all_customers to static method

JCantu248 committed:
Added functions to interact with UserReport Document Collections

Added functions to pca-import and pca-tool to write and read user report data from the PCA mongoDB. Also updated the database-model to no longer include method to find a UserReport from the DB, as it was no longer required as the pca-tool uses raw queries to retrieve data.

JCantu248 committed:
Bug fix for PCA-Import User Reports.

 Making sure import user_reports throws the correct error message when trying to import previously imported data.

dav3r commented on pca/db/database_model.py:
```diff
@@ -417,3 +419,36 @@ def save(self, *args, **kwargs):
+    assessment = fields.ReferenceField(AssessmentDoc, required=True)
+    campaign = fields.ReferenceField(CampaignDoc, required=True)
+    first_report = fields.DateTimeField(blank=True)
+    total_num_reports = fields.IntegerField(blank=True)
```
When would `total_num_reports` ever be blank?  When one of these documents is initially created, shouldn't that field be initialized to either zero or one (depending on when you plan to create these docs)?

JCantu248 replied:
You are correct. Its been changed to blank=False

dav3r commented on bin/pca-import:
```diff
@@ -773,6 +777,48 @@ def import_template_file(template_filename, template_id=None, force=False):
+                                                request["assessment"],
+                                                request["campaign"]):
+        print("Cannot import, user_report already exists: {} {} {}".format(
+            request["customer"], request["assessment"], request["campaign"]))
```
This error message text should probably make it clearer to the end user that the values being output correspond to `customer_id`, `assessment_id`, and `campaign_id`.

dav3r commented on bin/pca-import:
```diff
@@ -170,27 +172,29 @@ def import_assessment(request, source, force=False):
 
 
-def import_json_stdin(force, customerImport, assessmentImport):
+def import_json_stdin(force, customer_import, assessment_import):
```
This function also needs to be updated to handle `user_report_import`.  You should also test that this works as intended by passing in the user report JSON via `stdin`.

JCantu248 replied:
Tested with stdin. It is capable of adding a document to the database by console input

dav3r replied:
Looks good, thanks!

dav3r commented on bin/pca-tool:
```diff
@@ -419,6 +420,21 @@ def delete_image(image_id):
     return True
 
 
+def output_user_report(customer_id, assessment_id):
```
Since this function can output more than one user report, I'd call it `output_user_reports`:
```suggestion
def output_user_reports(customer_id, assessment_id):
```

Don't forget to update any other references to this function accordingly.

dav3r commented on pca/db/database_model.py:
```diff
@@ -417,3 +419,22 @@ def save(self, *args, **kwargs):
         super(ApplicationDoc, self).save(*args, **kwargs)
+
+class UserReportDoc(RootDoc):
+    """Provides schema for User Report doc."""
```
I'd prefer to see a bit more explanation in here about what this collection and the `first_report` date actually represent, since it's not clear from the code.

dav3r replied:
Looking good now after https://github.com/cisagov/pca-core/pull/5/commits/3733c13054a3c9a1a54dbb46983c9070043c733e and https://github.com/cisagov/pca-core/pull/5/commits/76ffea3eb11ba5c77d2593759504e71da3353c4e

dav3r commented on bin/pca-tool:
```diff
@@ -419,6 +420,21 @@ def delete_image(image_id):
+    user_reports = list(UserReportDoc.objects.raw({"customer": customer_id,
+                                                   "assessment": assessment_id}).order_by([("_id", pymongo.ASCENDING)]))
+    if user_reports:
+        print("customer_id", "campaign_id", "assessment_id", "first_report", "total_num_reports")
+        for user_report in user_reports:
+            print('"{0.customer._id}", "{0.assessment._id}", "{0.campaign._id}", {0.first_report}, {0.total_num_reports}'.format(
+                user_report
+            ))
```
The order of your field headings does not match the order of the data you are printing out (campaign and assessment IDs are reversed).

dav3r commented on bin/pca-tool:
```diff
@@ -419,6 +420,21 @@ def delete_image(image_id):
+            ))
+        return True
+    else:
+        print("No User Reports for customer {} and assessment {}".format(customer_id, assessment_id))
```
No need to capitalize here:
```suggestion
        print("No user reports for customer {} and assessment {}".format(customer_id, assessment_id))
```

dav3r reviewed (changes requested):
@JCantu248 - This is looking good, but please take a look at my comments and the changes I requested.

Also, please update the description of this PR to mention that you also added new options to both `pca-import` and `pca-tool`.  Even though you mentioned that in the testing, it's helpful to have the PR description contain a concise overview of the PR changes.

dav3r commented:
One other thing @JCantu248 - please assign yourself and add the `improvement` label to this PR.  Thanks!

JCantu248 committed:
Changes according to feedback.

Changes were made in accordance to feedback provided during code review. Also refactored the code to be able to handle overwriting previously entered data when using the --force flag.

dav3r commented:
@JCantu248 - In the future, please let the person who initiated each conversation be the one who marks the conversation as "Resolved".  That is how we typically know which items have not yet been resolved to our satisfaction.

Also, it is very helpful if you reply to each conversation with the commit URL where the issue was corrected (e.g. https://github.com/cisagov/pca-core/pull/5/commits/3733c13054a3c9a1a54dbb46983c9070043c733e).  Related to that, we prefer multiple, smaller commits, where each commit addresses a particular issue, rather than a single large commit that covers a wide variety of issues.  These two items combined make it much easier for a reviewer to re-review any updates that you made.

dav3r commented on bin/pca-import:
```diff
@@ -773,6 +779,59 @@ def import_template_file(template_filename, template_id=None, force=False):
+            user_report_id = user_report._id
+        else:
+            print(
+                "Cannot import, user_report already exist - Customer: {} Assessment: {} Campaign: {}".format(
```
Fix grammar and match other error messages:
```suggestion
                "Cannot import, user_report already exists - customer_id: {} assessment_id: {} campaign_id: {}".format(
```

JCantu248 replied:
Suggestion was added in 86972531a8facb739a161eecba3a8c1e546aca54

dav3r commented on pca/db/database_model.py:
```diff
@@ -417,3 +419,38 @@ def save(self, *args, **kwargs):
+    customer - a customer ID that corresponds to a CustomerDoc
+    assessment - an assessment ID that corresponds to an AssessmentDoc
+    campaign - a campaign ID that corresponds to a CampaignDoc
+    first_report - the first time a user responds to a phishing email. Optional, may not be available from GoPhish
+    total_num_reports - the total number of responses (reports) for a campaign.
```
How about this?  I don't think you need the "Optional..." bit, since that is covered by the field definition below.
```suggestion
    first_report - the timestamp when the first user click ("report") was received during a phishing campaign
    total_num_reports - the total number of user clicks ("reports") for a campaign
```

JCantu248 replied:
Changes applied in 76ffea3eb11ba5c77d2593759504e71da3353c4e

JCantu248 committed:
Update bin/pca-import

Committing suggestion to fix grammer from 'exist' to 'exists'

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

JCantu248 committed:
Update pca/db/database_model.py

Changing comments according to feedback

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r reviewed (approved):
I think this PR is good to go now - thanks for making the requested changes.  👍 

The one last thing that you should do is to delete the checklist item (in the PR description) that doesn't apply:  "Tests have been added and/or modified to cover the changes in this PR."

jsf9k commented on bin/pca-import:
```diff
@@ -170,27 +172,31 @@ def import_assessment(request, source, force=False):
 
 
-def import_json_file(filename, force, customerImport, assessmentImport):
+def import_json_file(filename, force, customer_import, assessment_import, user_report_import):
```
Do you intend this function to perform only one type of import?  What if someone passes `customer_import = True` and `assessment_import = True`?  I see that the code will only perform the customer import, but this may not be what the user expects.

The best way to address this would be for the function to accept a single `enum` value (`enums` are backported to Python 2.7 via [this `pip` package](https://pypi.org/project/enum34/)) value instead of the three Boolean values.  That said, I'm willing to approve if you create a docstring for this function describing that it expects _exactly one_ of the three Booleans to be `True`.

JCantu248 replied:
See a8014ba675ba14a104152964cb72cf7ce0bb7127 for my changes. I made it an enum value instead. It cleaned up the code somewhat

jsf9k commented on bin/pca-import:
```diff
@@ -170,27 +172,31 @@ def import_assessment(request, source, force=False):
 
 
-def import_json_stdin(force, customerImport, assessmentImport):
+def import_json_stdin(force, customer_import, assessment_import, user_report_import):
```
The comment above about `import_json_file()` also applies here.

JCantu248 replied:
See a8014ba675ba14a104152964cb72cf7ce0bb7127

jsf9k commented on pca/db/database_model.py:
```diff
@@ -417,3 +419,38 @@ def save(self, *args, **kwargs):
+
+        collection_name = USER_REPORT_COLLECTION
+        final = True
+        ignore_unknown_fields = True  # TODO see if this can be inherited from RootDoc
```
You should create an issue for this TODO, and include a link to the issue in this comment.

JCantu248 replied:
Issue can be found here: [Issue](https://github.com/cisagov/pca-core/issues/7)

jsf9k replied:
@JCantu248, please reference #7 in this comment.  Then re-request my review and I'll approve.

JCantu248 replied:
Umm okay.  #7

JCantu248 replied:
See commit abaedb5bdfa316bbf3ba6c0d93c873b56b431977 for code changes

jsf9k reviewed (changes requested):
Strong work @JCantu248!

I made a few minor requests.  Please re-request my review after you address them.

JCantu248 committed:
Changing import_json_file, import_json_stdin to use enum values.

The parameter list for import_json functions was using three separate boolean values, now it is using an enum value.

JCantu248 committed:
Merge branch 'add_user_report_doc' of github.com:cisagov/pca-core into add_user_report_doc

jsf9k reviewed (changes requested):
There is still one outstanding issue that should be resolved, then I can approve this PR.

dav3r commented on pca/db/database_model.py:
```diff
@@ -417,3 +419,38 @@ def save(self, *args, **kwargs):
+    assessment = fields.ReferenceField(AssessmentDoc, required=True)
+    campaign = fields.ReferenceField(CampaignDoc, required=True)
+    first_report = fields.DateTimeField(blank=True)
+    total_num_reports = fields.IntegerField(blank=False)
```
@JCantu248 - Sorry to go back to this, but is there any reason why this can't be:
```suggestion
    total_num_reports = fields.IntegerField(required=True)
```

JCantu248 replied:
I set it that way because I wanted to account for having no reports generated, but then I have first_report set to true, so they are contradictory. I may want to make first_report false as well in the event that no user responds to a phishing email.

JCantu248 committed:
Update pca/db/database_model.py

Change total_num_reports required=True

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

JCantu248 committed:
Adding issue #7 to comment on TODO for ignore_unknown_fields

jsf9k reviewed (approved):
Thanks for taking care of all those changes @JCantu248!
